### PR TITLE
Modify documentation of iterateN to be less confusing

### LINKED ISSUE #303: iterateN's documentation disagrees with its behaviour
The [documentation][1] for ``iterateN`` states:

> Apply function n times to value.

However, its behaviour disagrees with this, as the function is applied _max{n - 1, 0}_ times, not _n_ times:

```
> import Data.Vector as V
V> V.iterateN 5 succ (0 :: Int)
[0,1,2,3,4]
V> V.iterateN 0 succ (0 :: Int)
[]
V> V.iterateN 1 succ (0 :: Int)
[0]
```

This is clearly incorrect documentation, and is actually quite confusing - its description suggests the length of the resulting vector should be _n + 1_.

[1]: http://hackage.haskell.org/package/vector-0.12.1.2/docs/Data-Vector.html#v:iterateN

### PR DESCRIPTION
**Pull request by kozross**
Fixes #303. The docs now specify _both_ how many times the function is applied _and_ what the resulting length will be.

### CODE REVIEW

**Comment by Kleidukos** [comment]
@kozross I can't remember from the top of my head but doesn't `iterateNM` suffer from the same problem?

**Comment by kozross** [comment]
@Kleidukos Yes:

```
*Data.Vector.Internal.Check> import Data.Vector
*Data.Vector.Internal.Check Data.Vector> iterateNM 4 (\x -> Just (x + 1)) (0 :: Int)
Just [0,1,2,3]
*Data.Vector.Internal.Check Data.Vector> iterateNM 0 (\x -> Just (x + 1)) (0 :: Int)
Just []
*Data.Vector.Internal.Check Data.Vector> iterateNM 1 (\x -> Just (x + 1)) (0 :: Int)
Just [0]
```
It's arguably worse here, as you're in the context of a more-or-less arbitrary effect. Will amend their documentation also.

**Comment by cartazio** [comment]
i'm gonna squash and merge this into master today or tomorrow 

`Modify documentation of iterateN to clarify n-1 iterations plus initial value semantic (#304)`
is my tentative top of  commit title

**Comment by Shimuuar** [comment]
Thanks a lot! Could you update documentation in `Data.Vector.Fusion.*` as well?

**Comment by cartazio** [comment]
good catch

**Comment by Shimuuar** [comment]
@kozross Could you please update documentation in Data.Vector.Fusion.* as well. It' only thing preventing merge?

**Comment by Shimuuar** [comment]
Thank you! Merged

File: Data/Vector.hs [hunk 1]
```diff
@@ -667,7 +667,8 @@ generate :: Int -> (Int -> a) -> Vector a
 {-# INLINE generate #-}
 generate = G.generate
 
--- | /O(n)/ Apply function n times to value. Zeroth element is original value.
+-- | /O(n)/ Apply function \(\max\{n - 1, 0\}\) times to value, producing a 
+-- vector of length /n/. Zeroth element is original value.
```
[the remark this answers is not in the document]
Comment (kozross) [reply Data/Vector.hs:671]: That's actually a good point. I should fix that.

File: Data/Vector/Fusion/Bundle.hs [hunk 2]
```diff
@@ -442,7 +442,8 @@ unfoldrExactN :: Int -> (s -> (a, s)) -> s -> Bundle v a
 {-# INLINE unfoldrExactN #-}
 unfoldrExactN = M.unfoldrExactN
 
--- | Apply function n-1 times to value. Zeroth element is original value.
+-- | Apply function \(\max\{n - 1, 0\}\) times to a value, producing a vector of
+-- length /n/. Zeroth element is the original value.
```
[the remark this answers is not in the document]
Comment (kozross) [reply Data/Vector/Fusion/Bundle.hs:446]: Where exactly did these show up? I can adjust (or give justifications), but I had to fix this in like seven places.

**Comment by kozross** [comment]
@Bodigrim I am happy to represent it that way - it's clearer than my text.

File: Data/Vector/Fusion/Bundle.hs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Bodigrim) [reply Data/Vector/Fusion/Bundle.hs:446]: `D.V`, `D.V.Generic`, `D.V.Primitive`, `D.V.Storable`, `D.V.Unboxed` - you left "to value" and "is original value" as is.

`D.V.Fusion.Bundle.iterateN` - you added articles in both places.
`D.V.Fusion.Bundle.Monadic.iterateNM` - you left "to value", but added "is **the** original value".
`D.V.Fusion.Bundle.Monadic.iterateN` - you added articles in both places.
`D.V.Fusion.Stream.Monadic.iterateNM` - you left "to value", but added "is **the** original value".
`D.V.Fusion.Stream.Monadic.iterateN` - you added articles in both places.

**Comment by Bodigrim** [comment]
It would be nice to fix at least https://github.com/haskell/vector/pull/304#discussion_r437580490 before 0.12.2.0 release.
CC @lehins @kozross

**Comment by lehins** [comment]
@Bodigrim Yeah, that's a good idea. I'll get it done

**Comment by lehins** [comment]
Here is a PR with this fix: #365

**Pull request merged**